web/packages/teleport/src/Discover/Shared/useJoinToken.ts

```diff
@@ -0,0 +1,101 @@
+import { useEffect, useState } from 'react';
+
+import { useTeleport } from 'teleport';
+
+import {
+  ResourceKind,
+  resourceKindToJoinRole,
+} from 'teleport/Discover/Shared/ResourceKind';
+
+import type { AgentLabel } from 'teleport/services/agents';
+import type { JoinMethod, JoinToken } from 'teleport/services/joinToken';
+
+interface SuspendResult {
+  promise?: Promise<any>;
+  response?: JoinToken;
+  error?: Error;
+}
+
+let abortController: AbortController;
+let joinTokenCache = new Map<ResourceKind, SuspendResult>();
+
+export function clearCachedJoinTokenResult(resourceKind: ResourceKind) {
+  joinTokenCache.delete(resourceKind);
+}
+
+export function useJoinToken(
```

Comment:
Thinking that we should create (and document in a developer section of the README) a naming convention that denotes this as a suspense hook (perhaps something like `useJoinTokenSuspender`). 

My reasoning is that `Suspense` is an experimental feature, and it's API is still unconfirmed and I wasn't able to find reliable documentation for it when initially googling. If I was new to the project, I would be totally baffled by this component.

That [example you linked me](https://codesandbox.io/s/frosty-hermann-bztrp?file=/src/fakeApi.js) is pretty good, although I wonder if there's also an official or unofficial documentation page that I missed.

Reply:
Sure, we can rename this